File: src/live/candle_feed.py
"""Provides 5m candle data for live trading."""

from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from typing import Optional

import structlog

from ..core.models.candle import Candle, CandleSeries

logger = structlog.get_logger(__name__)


def _ohlcv_to_candle(row: list, symbol: str, timeframe: str) -> Candle:
    """Convert a ccxt OHLCV row [ts_ms, o, h, l, c, v] to a Candle."""
    ts_ms, o, h, l, c, v = row[0], row[1], row[2], row[3], row[4], row[5]
    return Candle(
        symbol=symbol,
        exchange="bybit",
        timeframe=timeframe,
        timestamp=datetime.fromtimestamp(ts_ms / 1000, tz=timezone.utc),
        open=o,
        high=h,
        low=l,
        close=c,
        volume=v,
    )


class CandleFeed:
    """Provides 5m candle data for live trading.

    Uses ExchangeClient (ccxt) for REST OHLCV fetching.
    Maintains a rolling window of candles per symbol for signal generation.
    Supports resume from last known timestamp via StateManager.
    """

    def __init__(
        self,
        exchange_client,  # ExchangeClient — ccxt wrapper
        state_manager,    # StateManager — resume timestamp
        symbols: list[str],
        timeframe: str = "5m",
        lookback_candles: int = 200,
        market_type: str = "perp",
    ) -> None:
        self._exchange = exchange_client
        self._state = state_manager
        self._symbols = symbols
        self._timeframe = timeframe
        self._lookback = lookback_candles
        self._market_type = market_type
        self._buffers: dict[str, list[Candle]] = {}
        self._series: dict[str, CandleSeries] = {}
        self._log = logger.bind(component="candle_feed")

    # ── lifecycle ──────────────────────────────────────────────

    async def initialize(self) -> None:
        """Load historical candles for each symbol.

        For each symbol:
        1. Check state_manager for a resume timestamp
        2. If found, fetch candles from that timestamp onward
        3. Otherwise, fetch the last lookback_candles
        4. Build per-symbol CandleSeries
        """
        for symbol in self._symbols:
            try:
                candles = await self._load_historical(symbol)
                self._buffers[symbol] = candles
                self._series[symbol] = self._build_series(symbol, candles)
                self._log.info(
                    "initialized",
                    symbol=symbol,
                    candles=len(candles),
                )
            except Exception as exc:
                self._log.warning("init_failed", symbol=symbol, error=str(exc))
                self._buffers[symbol] = []
                self._series[symbol] = self._build_series(symbol, [])

    async def _load_historical(self, symbol: str) -> list[Candle]:
        """Fetch historical candles, resuming from last known timestamp."""
        resume_ts = await self._state.get_state(f"last_candle_{symbol}")

        if resume_ts:
            since = int(
                datetime.fromisoformat(resume_ts).timestamp() * 1000
            )
            rows = await self._fetch_ohlcv(symbol, since=since)
        else:
            rows = await self._fetch_ohlcv(symbol)

        return [_ohlcv_to_candle(r, symbol, self._timeframe) for r in rows]

    async def _fetch_ohlcv(
        self,
        symbol: str,
        since: Optional[int] = None,
    ) -> list[list]:
        """Call ccxt fetch_ohlcv through the ExchangeClient."""
        exchange = self._exchange._exchange  # ccxt instance
        ccxt_symbol = _normalize(symbol, self._market_type)
        return await exchange.fetch_ohlcv(
            ccxt_symbol,
            self._timeframe,
            since=since,
            limit=self._lookback,
        )

    @staticmethod
    def _normalize(symbol: str, market_type: str = "perp") -> str:
        """Convert raw symbol to ccxt format: BTCUSDT -> BTC/USDT:USDT."""
        clean = symbol.upper().strip()
        if ":" in clean:
            return clean
        base = clean.split("/")[0] if "/" in clean else clean.rstrip("USDT")
        if market_type == "perp":
            return f"{base}/USDT:USDT"
        return f"{base}/USDT"

    # ── live updates ───────────────────────────────────────────

    async def fetch_latest_candle(self, symbol: str) -> Optional[Candle]:
        """Fetch the most recent completed candle for a symbol.

        Fetches the last 2 OHLCV rows; the second-to-last is the
        last *completed* candle (the final row may be still forming).
        Returns None on error (logs warning, never raises).
        """
        try:
            exchange = self._exchange._exchange
            ccxt_symbol = self._normalize(symbol, self._market_type)
            rows = await exchange.fetch_ohlcv(
                ccxt_symbol, self._timeframe, limit=2,
            )
            if not rows:
                return None
            # Use second-to-last (last fully closed candle)
            idx = -2 if len(rows) >= 2 else -1
            return _ohlcv_to_candle(rows[idx], symbol, self._timeframe)
        except Exception as exc:
            self._log.warning("fetch_failed", symbol=symbol, error=str(exc))
            return None

    async def update_candles(self, symbol: str) -> Optional[CandleSeries]:
        """Fetch latest candle and update the rolling buffer if new.

        1. Fetch latest completed candle
        2. If its timestamp is newer than the buffer tail, append
        3. Trim to lookback window, persist timestamp, rebuild series
        4. Return updated CandleSeries or None if no new data
        """
        candle = await self.fetch_latest_candle(symbol)
        if candle is None:
            return None

        buf = self._buffers.get(symbol, [])
        if buf and candle.timestamp <= buf[-1].timestamp:
            return None  # No new candle yet

        updated = [*buf, candle][-self._lookback :]
        self._buffers[symbol] = updated
        series = self._build_series(symbol, updated)
        self._series[symbol] = series

        await self._state.set_state(
            f"last_candle_{symbol}", candle.timestamp.isoformat()
        )
        self._log.debug(
            "candle_updated",
            symbol=symbol,
            ts=candle.timestamp.isoformat(),
        )
        return series

    def get_candle_series(self, symbol: str) -> Optional[CandleSeries]:
        """Get current CandleSeries for a symbol. None if not initialized."""
        return self._series.get(symbol)

    async def wait_for_candle(
        self, symbol: str, timeout: float = 310,
        shutdown_event: Optional[asyncio.Event] = None,
    ) -> Optional[CandleSeries]:
        """Poll every 10s until a new candle arrives or timeout.

        This is the main entry point for the trading loop — it blocks
        until fresh candle data is available.
        """
        poll_interval = 10
        elapsed = 0.0

        while elapsed < timeout:
            if shutdown_event and shutdown_event.is_set():
                return None
            series = await self.update_candles(symbol)
            if series is not None:
                return series
            try:
                await asyncio.wait_for(
                    shutdown_event.wait() if shutdown_event else asyncio.sleep(poll_interval),
                    timeout=poll_interval,
                )
                # Event was set → shutting down
                return None
            except asyncio.TimeoutError:
                pass
            elapsed += poll_interval

        self._log.warning("wait_timeout", symbol=symbol, timeout=timeout)
        return None

    # ── helpers ────────────────────────────────────────────────

    @staticmethod
    def _build_series(symbol: str, candles: list[Candle]) -> CandleSeries:
        """Construct a CandleSeries from a candle list."""
        return CandleSeries(
            candles=candles,
            symbol=symbol,
            exchange="bybit",
            timeframe="5m",
        )

File: pypsik-live/src/live/candle_feed.rs
/// Candle feed — REST polling for OHLCV data with rolling buffers.

use std::collections::HashMap;

use tokio::sync::broadcast;
use tracing::{debug, info, warn};

use crate::exchange::client::ExchangeClient;
use crate::live::state_manager::StateManager;
use crate::models::candle::{ohlcv_to_candle, Candle, CandleSeries};

/// Provides candle data for live trading via REST polling.
pub struct CandleFeed {
    exchange: ExchangeClient,
    state: StateManager,
    symbols: Vec<String>,
    timeframe: String,
    lookback: usize,
    market_type: String,
    buffers: HashMap<String, Vec<Candle>>,
    series: HashMap<String, CandleSeries>,
}

impl CandleFeed {
    pub fn new(
        exchange: ExchangeClient,
        state: StateManager,
        symbols: &[String],
        timeframe: &str,
        lookback: usize,
        market_type: &str,
    ) -> Self {
        Self {
            exchange,
            state,
            symbols: symbols.to_vec(),
            timeframe: timeframe.to_string(),
            lookback,
            market_type: market_type.to_string(),
            buffers: HashMap::new(),
            series: HashMap::new(),
        }
    }

    /// Load historical candles for each symbol.
    pub async fn initialize(&mut self) {
        for symbol in &self.symbols {
            match self.load_historical(symbol).await {
                Ok(candles) => {
                    info!(symbol = %symbol, candles = candles.len(), "candle_feed_initialized");
                    let series = self.build_series(symbol, &candles);
                    self.buffers.insert(symbol.clone(), candles);
                    self.series.insert(symbol.clone(), series);
                }
                Err(e) => {
                    warn!(symbol = %symbol, error = %e, "candle_feed_init_failed");
                    let empty = Vec::new();
                    let series = self.build_series(symbol, &empty);
                    self.buffers.insert(symbol.clone(), empty);
                    self.series.insert(symbol.clone(), series);
                }
            }
        }
    }

    async fn load_historical(&self, symbol: &str) -> Result<Vec<Candle>, Box<dyn std::error::Error + Send + Sync>> {
        let since = match self.state.get_state(&format!("last_candle_{}", symbol)).await? {
            Some(ts) => {
                let dt = chrono::DateTime::parse_from_rfc3339(&ts)?;
                Some(dt.timestamp_millis())
            }
            None => None,
        };

        let rows = self.exchange.fetch_ohlcv(symbol, &self.timeframe, since, Some(self.lookback as u32)).await?;
        let candles: Vec<Candle> = rows
            .iter()
            .map(|r| ohlcv_to_candle(r, symbol, &self.timeframe))
            .collect();
        Ok(candles)
    }

    /// Fetch latest completed candle and update the buffer.
    pub async fn update_candles(
        &mut self,
        symbol: &str,
    ) -> Option<CandleSeries> {
        let candle = match self.fetch_latest_candle(symbol).await {
            Some(c) => c,
            None => return None,
        };

        let buf = self.buffers.get(symbol)?;
        if !buf.is_empty() && candle.timestamp <= buf.last().unwrap().timestamp {
            return None; // No new candle
        }

        let mut updated = buf.clone();
        updated.push(candle.clone());
        if updated.len() > self.lookback {
            updated = updated.split_off(updated.len() - self.lookback);
        }

        let _ = self
            .state
            .set_state(
                &format!("last_candle_{}", symbol),
                &candle.timestamp.to_rfc3339(),
            )
            .await;

        let series = self.build_series(symbol, &updated);
        self.buffers.insert(symbol.to_string(), updated);
        self.series.insert(symbol.to_string(), series.clone());

        debug!(symbol = %symbol, ts = %candle.timestamp.to_rfc3339(), "candle_updated");
        Some(series)
    }

    /// Get current CandleSeries for a symbol.
    pub fn get_candle_series(&self, symbol: &str) -> Option<&CandleSeries> {
        self.series.get(symbol)
    }

    /// Poll every 10s until a new candle arrives, timeout, or shutdown.
    pub async fn wait_for_candle(
        &mut self,
        symbol: &str,
        shutdown: &mut broadcast::Receiver<()>,
        timeout_secs: u64,
    ) -> Option<CandleSeries> {
        let poll_interval = std::time::Duration::from_secs(10);
        let start = std::time::Instant::now();
        let timeout = std::time::Duration::from_secs(timeout_secs);

        loop {
            // Non-blocking shutdown check
            match shutdown.try_recv() {
                Ok(_) => return None,
                Err(tokio::sync::broadcast::error::TryRecvError::Empty) => {} // continue
                Err(tokio::sync::broadcast::error::TryRecvError::Closed) => return None,
                Err(tokio::sync::broadcast::error::TryRecvError::Lagged(_)) => {} // ignore lag
            }

            if let Some(series) = self.update_candles(symbol).await {
                return Some(series);
            }

            if start.elapsed() >= timeout {
                warn!(symbol = %symbol, timeout_secs, "wait_timeout");
                return None;
            }

            // Sleep with shutdown awareness
            tokio::select! {
                _ = tokio::time::sleep(poll_interval) => {},
                _ = shutdown.recv() => return None,
            }
        }
    }

    /// Fetch the most recent completed candle.
    async fn fetch_latest_candle(&self, symbol: &str) -> Option<Candle> {
        match self.exchange.fetch_ohlcv(symbol, &self.timeframe, None, Some(2)).await {
            Ok(rows) => {
                if rows.is_empty() {
                    return None;
                }
                // Use second-to-last (last fully closed candle)
                let idx = if rows.len() >= 2 { rows.len() - 2 } else { 0 };
                Some(ohlcv_to_candle(&rows[idx], symbol, &self.timeframe))
            }
            Err(e) => {
                warn!(symbol = %symbol, error = %e, "fetch_latest_failed");
                None
            }
        }
    }

    fn build_series(&self, symbol: &str, candles: &[Candle]) -> CandleSeries {
        CandleSeries {
            candles: candles.to_vec(),
            symbol: symbol.to_string(),
            exchange: "bybit".to_string(),
            timeframe: self.timeframe.clone(),
        }
    }
}
